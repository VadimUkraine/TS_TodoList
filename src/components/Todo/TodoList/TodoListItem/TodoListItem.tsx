import React, { useState, useEffect, useRef } from "react";
import ListItem from "@material-ui/core/ListItem";
import ListItemSecondaryAction from "@material-ui/core/ListItemSecondaryAction";
import ListItemText from "@material-ui/core/ListItemText";
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import TextField from "@material-ui/core/TextField";
import { createStyles, makeStyles } from "@material-ui/core/styles";
import { Todo } from '../../../../types';

type Props = {
  todo: Todo;
  deleteTodo: (todo: string) => void;
  editID: string;
  setEditId: (id: string) => void;
  changeTodo: (id: string, todo: string) => void;
}

const useStyles = makeStyles(() => createStyles({
  todoItemContent: {
    display: "flex",
    flexDirection: "column",
    minWidth: "100%",
    height: "2.8125rem",
  },
  todoButtonsWrap: {
    right: "0rem",
  },
  btnDelete: {
    marginLeft: "0.3125rem",
  },
  todoItem: {
    paddingLeft: "0rem",
    paddingRight: "9.375rem",
    paddingTop: "0.9375rem",
  },
  textStyles: {
    width: "100%",
    whiteSpace: "nowrap",
    overflow: "hidden",
    textOverflow: "ellipsis",
    lineHeight: "1rem",
  },
  dateStyles: {
    textAlign: "right",
    fontWeight: 600,
    opacity: 0.5,
  },
  editInputStyles: {
    padding: "0.125rem 0rem 0.1875rem 0rem",
  },
}));

export const TodoListItem: React.FC<Props> = ({
  todo, deleteTodo, editID, setEditId, changeTodo,
}) => {

  const [inputValue, setInputValue] = useState('');
  const refInput = useRef(null);

  useEffect(() => {
    if (editID === todo.id && refInput.current) {
      refInput.current.focus();
    }
  }, [editID, todo]);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setInputValue(event.target.value);
  };

  const handleDelete = () => {
    deleteTodo(todo.id);
  };

  const handleEdit = () => {
    setEditId(todo.id);
    setInputValue(todo.text);
  };

  const handleBlur = () => {
    setEditId("");
  };

  const handleKeyPress = (event: any) => {
    if (event.key !== 'Enter') return;
    if (inputValue !== todo.text) {
      changeTodo(todo.id, inputValue);
    }
    handleBlur();
  };

  const classes = useStyles();

  return (
      <ListItem
      className={classes.todoItem}
      data-testid={"todo-list-item"}
      >
        <div className={classes.todoItemContent}>
          {editID !== todo.id && <ListItemText
            primary={todo.text}
            classes={{
              primary: classes.textStyles,
            }}
          />}
          {editID === todo.id && <TextField
            value={inputValue}
            onChange={handleChange}
            onBlur={handleBlur}
            onKeyPress={handleKeyPress}
            inputProps={{
              "aria-label": "todo-item-edit-input",
              className: classes.editInputStyles,
              ref: refInput,
            }}
        />}
        <Typography className={classes.dateStyles} variant="caption" display="block" gutterBottom>
          {todo.date}
        </Typography>
        </div>
        <ListItemSecondaryAction
          className={classes.todoButtonsWrap}
        >
          <Button size="small" variant="contained" color="primary" onClick={handleEdit}>
            Edit
          </Button>
          <Button className={classes.btnDelete} size="small" variant="contained" color="secondary" onClick={handleDelete}>
            Delete
          </Button>
        </ListItemSecondaryAction>
      </ListItem>
  );
};

export default TodoListItem;
