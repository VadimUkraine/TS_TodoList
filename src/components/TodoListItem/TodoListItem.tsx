import React from "react";
import ListItem from "@material-ui/core/ListItem";
import ListItemSecondaryAction from "@material-ui/core/ListItemSecondaryAction";
import ListItemText from "@material-ui/core/ListItemText";
import IconButton from "@material-ui/core/IconButton";
import DeleteIcon from "@material-ui/icons/Delete";

interface IProps {
  todo: string;
  deleteTodo: (todo: string)=> void;
}

export const TodoListItem: React.FC<IProps> = ({ todo, deleteTodo }) => {

  const handleDelete = () => {
    deleteTodo(todo);
  };

  return (
      <>
        <ListItem>
          <ListItemText
            primary={todo}
          />
          <ListItemSecondaryAction>
            <IconButton edge="end" aria-label="delete" onClick={handleDelete}>
              <DeleteIcon />
            </IconButton>
          </ListItemSecondaryAction>
        </ListItem>
      </>
  );
};

export default TodoListItem;
