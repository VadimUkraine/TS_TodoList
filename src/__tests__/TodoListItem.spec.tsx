import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import {
  render, screen, fireEvent, cleanup,
} from '@testing-library/react';
import TodoListItem from '../components/Todo/TodoList/TodoListItem';

const mockTodoItem = {
  id: 'a52dfc2d-0d70-49b9-b58f-21b8ea2408d4',
  text: 'buy milk',
  date: '03.11.2020, 20:38',
};
const handleDelete = jest.fn();
const handleChangeTodo = jest.fn();
const handleSetEditId = jest.fn();

beforeEach(() => render(
  <TodoListItem
    todo={mockTodoItem}
    deleteTodo={handleDelete}
    editID={mockTodoItem.id}
    setEditId={handleSetEditId}
    changeTodo={handleChangeTodo}
  />,
));

afterEach(cleanup);

test('it renders component', () => {
  expect(screen.getByRole('textbox', { name: 'todo-item-edit-input' })).toBeInTheDocument();
});

test('it renders delete button', () => {
  expect(screen.getByRole('button', { name: 'Delete' })).toBeInTheDocument();
});

test('it renders edit button', () => {
  expect(screen.getByRole('button', { name: 'Edit' })).toBeInTheDocument();
});

test('it checks delete item from DOM', () => {
  fireEvent.click(screen.getByRole('button', { name: 'Delete' }));
  expect(screen.queryByText(/buy milk/i)).toBeNull();
});

test('it checks click button edit', () => {
  fireEvent.click(screen.getByRole('button', { name: 'Edit' }));
  expect(screen.getByDisplayValue('buy milk')).toBeInTheDocument();
});

test('it checks click delete button', () => {
  fireEvent.click(screen.getByRole('button', { name: 'Delete' }));
  expect(handleDelete).toBeCalled();
});

test('it renders correctly todoListItem', () => {
  const tree = renderer
    .create(<TodoListItem
      todo={mockTodoItem}
      deleteTodo={handleDelete}
      editID={mockTodoItem.id}
      setEditId={handleSetEditId}
      changeTodo={handleChangeTodo}
    />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
