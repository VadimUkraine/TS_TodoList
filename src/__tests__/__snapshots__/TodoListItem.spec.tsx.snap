// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly todoListItem 1`] = `
<li
  className="MuiListItem-container"
>
  <div
    className="MuiListItem-root makeStyles-todoItem-46 MuiListItem-gutters MuiListItem-secondaryAction"
    data-testid="todo-list-item"
    disabled={false}
  >
    <div
      className="makeStyles-todoItemContent-43"
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
        onKeyPress={[Function]}
      >
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            aria-label="todo-item-edit-input"
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input makeStyles-editInputStyles-49"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
        </div>
      </div>
      <span
        className="MuiTypography-root makeStyles-dateStyles-48 MuiTypography-caption MuiTypography-gutterBottom MuiTypography-displayBlock"
      >
        03.11.2020, 20:38
      </span>
    </div>
  </div>
  <div
    className="MuiListItemSecondaryAction-root makeStyles-todoButtonsWrap-44"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Edit
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-btnDelete-45 MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Delete
      </span>
    </button>
  </div>
</li>
`;
